{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7c312bf2",
   "metadata": {},
   "source": [
    "We will perform dNdS using probabilities for single base substitutions (single nucleotide context)\n",
    "\n",
    "Do we have patients with 2 mutations in a single gene?\n",
    "\n",
    "General steps:\n",
    "1. Get the probabilities for all 6 types of single base substitutions\n",
    "   - C>A\n",
    "   - C>G\n",
    "   - C>T\n",
    "   - T>A\n",
    "   - T>C\n",
    "   - T>G\n",
    "2. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "664ada56",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "\n",
    "# change working directory to project-2\n",
    "if os.getcwd().split('/')[-1] != 'project-2':\n",
    "    os.chdir('../../../')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "143ad330",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set paths\n",
    "METADATA = 'data/raw/TCGA.BRCA.metadata.txt'\n",
    "MUTATIONS = 'data/processed/TCGA.BRCA.mutations.qc1.txt'\n",
    "CDS_LENGTHS = 'data/processed/gencode.v23lift37.pc_transcripts.transcripts_in_TCGA_MAF.cds_lengths.tsv'\n",
    "dnds_opportunities = 'data/processed/dnds_opportunities.tsv'\n",
    "\n",
    "df_opportunities = pd.read_csv(dnds_opportunities, sep='\\t')\n",
    "df_mut = pd.read_csv(MUTATIONS, sep='\\t')\n",
    "df_meta = pd.read_csv(METADATA, sep='\\t')\n",
    "df_cds = pd.read_csv(CDS_LENGTHS, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "142f0790",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = \"data/processed/gencode.v23lift37.pc_transcripts.transcripts_in_TCGA_MAF.fa\"\n",
    "df = pd.read_csv(file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a11315bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>CDS_length</th>\n",
       "      <th>CDS sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1BG</td>\n",
       "      <td>ENST00000263100</td>\n",
       "      <td>1488</td>\n",
       "      <td>ATGTCCATGCTCGTGGTCTTTCTCTTGCTGTGGGGTGTCACCTGGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995</td>\n",
       "      <td>1785</td>\n",
       "      <td>ATGGAAGCAGTGTGTCTGGGCACATGCCCAGAGCCAGAAGCGAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2M</td>\n",
       "      <td>ENST00000318602</td>\n",
       "      <td>4425</td>\n",
       "      <td>ATGGGGAAGAACAAACTCCTTCATCCAAGTCTGGTTCTTCTCCTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2ML1</td>\n",
       "      <td>ENST00000299698</td>\n",
       "      <td>4365</td>\n",
       "      <td>ATGTGGGCTCAGCTCCTTCTAGGAATGTTGGCCCTATCACCAGCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>ENST00000401850</td>\n",
       "      <td>1062</td>\n",
       "      <td>ATGTCCAAGCCCCCCGACCTCCTGCTGCGGCTGCTCCGGGGCGCCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol    Transcript_ID  CDS_length  \\\n",
       "0        A1BG  ENST00000263100        1488   \n",
       "1        A1CF  ENST00000373995        1785   \n",
       "2         A2M  ENST00000318602        4425   \n",
       "3       A2ML1  ENST00000299698        4365   \n",
       "4      A4GALT  ENST00000401850        1062   \n",
       "\n",
       "                                        CDS sequence  \n",
       "0  ATGTCCATGCTCGTGGTCTTTCTCTTGCTGTGGGGTGTCACCTGGG...  \n",
       "1  ATGGAAGCAGTGTGTCTGGGCACATGCCCAGAGCCAGAAGCGAGCA...  \n",
       "2  ATGGGGAAGAACAAACTCCTTCATCCAAGTCTGGTTCTTCTCCTCT...  \n",
       "3  ATGTGGGCTCAGCTCCTTCTAGGAATGTTGGCCCTATCACCAGCCA...  \n",
       "4  ATGTCCAAGCCCCCCGACCTCCTGCTGCGGCTGCTCCGGGGCGCCC...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb938086",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "\n",
    "# Mutation count table: SBS_count[ref][alt]\n",
    "SBS_count = defaultdict(lambda: defaultdict(int))\n",
    "\n",
    "def normalize_sbs(ref, alt):\n",
    "    # Normalize purine ref bases to pyrimidine (G→C, A→T; and likewise alt)\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    if ref in ['G', 'A']:\n",
    "        ref = complement[ref]\n",
    "        alt = complement[alt]\n",
    "    return ref, alt\n",
    "\n",
    "for hgvsc in df_mut['HGVSc']:\n",
    "    m = re.search(r'c\\.\\d+([ACGT])>([ACGT])', str(hgvsc))\n",
    "    if not m:\n",
    "        continue\n",
    "    ref, alt = m.group(1), m.group(2)\n",
    "    norm_ref, norm_alt = normalize_sbs(ref, alt)\n",
    "    SBS_count[norm_ref][norm_alt] += 1\n",
    "\n",
    "records = []\n",
    "for ref in ['C', 'T']:\n",
    "    for alt in ['A', 'G', 'T'] if ref == 'C' else ['A', 'C', 'G']:\n",
    "        count = SBS_count[ref][alt]\n",
    "        records.append({'ref': ref, 'alt': alt, 'mutation': f\"{ref}>{alt}\", 'count': count})\n",
    "\n",
    "sbs_df = pd.DataFrame(records)\n",
    "\n",
    "# Count total number of 'C' and 'T' bases across all CDSs\n",
    "base_counts = defaultdict(int)\n",
    "\n",
    "for seq in df_cds['CDS sequence']:\n",
    "    seq = str(seq).upper()\n",
    "    base_counts['C'] += seq.count('C')\n",
    "    base_counts['T'] += seq.count('T')\n",
    "\n",
    "# Multiply by 2 * number of individuals (diploid genomes)\n",
    "num_individuals = len(df_meta)\n",
    "multiplier = 2 * num_individuals\n",
    "\n",
    "# Assign opportunities for each of the 6 SBS types\n",
    "opportunities = {\n",
    "    'C>A': base_counts['C'] * multiplier,\n",
    "    'C>G': base_counts['C'] * multiplier,\n",
    "    'C>T': base_counts['C'] * multiplier,\n",
    "    'T>A': base_counts['T'] * multiplier,\n",
    "    'T>C': base_counts['T'] * multiplier,\n",
    "    'T>G': base_counts['T'] * multiplier\n",
    "}\n",
    "\n",
    "sbs_df['opportunities'] = sbs_df['mutation'].map(opportunities)\n",
    "sbs_df['rate'] = sbs_df['count'] / sbs_df['opportunities']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d297023b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "from itertools import product\n",
    "\n",
    "# DNA codon table\n",
    "codon_table = {\n",
    "    'TTT': 'F', 'TTC': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "    'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S',\n",
    "    'TAT': 'Y', 'TAC': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "    'TGT': 'C', 'TGC': 'C', 'TGA': '*', 'TGG': 'W',\n",
    "\n",
    "    'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',\n",
    "    'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "    'CAT': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "\n",
    "    'ATT': 'I', 'ATC': 'I', 'ATA': 'I', 'ATG': 'M',\n",
    "    'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "    'AAT': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'AGT': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "\n",
    "    'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',\n",
    "    'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "    'GAT': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "}\n",
    "\n",
    "\n",
    "mutation_rate_map = sbs_df.set_index('mutation')['rate'].to_dict()\n",
    "\n",
    "codon_opportunity = {}\n",
    "\n",
    "for codon in codon_table:\n",
    "    syn = 0.0\n",
    "    nonsyn = 0.0\n",
    "    ref_aa = codon_table[codon]\n",
    "\n",
    "    for pos in range(3):\n",
    "        ref_base = codon[pos]\n",
    "        for alt_base in \"ACGT\":\n",
    "            if alt_base == ref_base:\n",
    "                continue\n",
    "            mut_codon = codon[:pos] + alt_base + codon[pos+1:]\n",
    "            if mut_codon not in codon_table:\n",
    "                continue\n",
    "            alt_aa = codon_table[mut_codon]\n",
    "\n",
    "            norm_ref, norm_alt = normalize_sbs(ref_base, alt_base)\n",
    "            key = f\"{norm_ref}>{norm_alt}\"\n",
    "            rate = mutation_rate_map.get(key, 0.0)\n",
    "\n",
    "            if ref_aa == alt_aa:\n",
    "                syn += rate\n",
    "            else:\n",
    "                nonsyn += rate\n",
    "\n",
    "    codon_opportunity[codon] = (syn, nonsyn)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f70699e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "\n",
    "for idx, row in df_cds.iterrows():\n",
    "    gene = row['gene']\n",
    "    seq = row['CDS sequence'].upper()\n",
    "\n",
    "    O_S = observed_synonymous.get(gene, 0)\n",
    "    O_NS = observed_nonsynonymous.get(gene, 0)\n",
    "\n",
    "    E_S, E_NS = 0.0, 0.0\n",
    "    for i in range(0, len(seq) - 2, 3):\n",
    "        codon = seq[i:i+3]\n",
    "        if codon not in codon_opportunity:\n",
    "            continue\n",
    "        syn, nonsyn = codon_opportunity[codon]\n",
    "        E_S += syn\n",
    "        E_NS += nonsyn\n",
    "\n",
    "    λ = O_S / E_S if E_S > 0 else None\n",
    "    expected_NS_scaled = λ * E_NS if λ is not None else None\n",
    "    dnds = O_NS / expected_NS_scaled if expected_NS_scaled else None\n",
    "\n",
    "    results.append({\n",
    "        'gene': gene,\n",
    "        'O_S': O_S,\n",
    "        'O_NS': O_NS,\n",
    "        'E_S': E_S,\n",
    "        'E_NS': E_NS,\n",
    "        'λ': λ,\n",
    "        'dN/dS': dnds\n",
    "    })\n",
    "\n",
    "df_dnds = pd.DataFrame(results)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "project2_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
