{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7c312bf2",
   "metadata": {},
   "source": [
    "We will perform dNdS using probabilities for single base substitutions (single nucleotide context)\n",
    "\n",
    "Do we have patients with 2 mutations in a single gene?\n",
    "\n",
    "General steps:\n",
    "1. Get the probabilities for all 6 types of single base substitutions\n",
    "   - C>A\n",
    "   - C>G\n",
    "   - C>T\n",
    "   - T>A\n",
    "   - T>C\n",
    "   - T>G\n",
    "2. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "664ada56",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import poisson\n",
    "import os\n",
    "\n",
    "import re\n",
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "from itertools import product\n",
    "\n",
    "# change working directory to project-2\n",
    "if os.getcwd().split('/')[-1] != 'project-2':\n",
    "    os.chdir('../../../')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "143ad330",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set paths\n",
    "METADATA = 'data/raw/TCGA.BRCA.metadata.txt'\n",
    "MUTATIONS = 'data/processed/TCGA.BRCA.mutations.qc1.txt'\n",
    "CDS_LENGTHS = 'data/processed/gencode.v23lift37.pc_transcripts.transcripts_in_TCGA_MAF.cds_lengths.tsv'\n",
    "dnds_opportunities = 'data/processed/dnds_opportunities.tsv'\n",
    "\n",
    "df_opportunities = pd.read_csv(dnds_opportunities, sep='\\t')\n",
    "df_mut = pd.read_csv(MUTATIONS, sep='\\t')\n",
    "df_meta = pd.read_csv(METADATA, sep='\\t')\n",
    "df_cds = pd.read_csv(CDS_LENGTHS, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "aaf0ae47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. Define categories and order\n",
    "syn_col = 'synonymous'\n",
    "non_syn_classes = [\n",
    "    'Frame_Shift_Del', \n",
    "    'Frame_Shift_Ins', \n",
    "    'In_Frame_Del', \n",
    "    'In_Frame_Ins',  \n",
    "    'Missense_Mutation', \n",
    "    'Nonsense_Mutation', \n",
    "    'Nonstop_Mutation',\n",
    "    'Translation_Start_Site'\n",
    "]\n",
    "all_classes = [syn_col] + non_syn_classes\n",
    "\n",
    "# 3. Group and pivot to get counts per gene per class\n",
    "mutation_counts = df_mut.groupby(['Hugo_Symbol', 'mutation_class']).size()\n",
    "counts_df = mutation_counts.unstack(fill_value=0)\n",
    "# 4. Ensure all classes exist\n",
    "for col in all_classes:\n",
    "    if col not in counts_df.columns:\n",
    "        counts_df[col] = 0\n",
    "\n",
    "# 5. Reorder columns\n",
    "counts_df = counts_df[all_classes]\n",
    "\n",
    "counts_df['non-synonymous'] = counts_df[non_syn_classes].sum(axis=1)\n",
    "\n",
    "# 6. Filter genes with at least 5 total mutations))\n",
    "counts_df = counts_df[counts_df.sum(axis=1) >= 5]\n",
    "\n",
    "# 7. Normalize by CDS length\n",
    "counts_df = counts_df.join(df_opportunities['CDS_length'], how='left')\n",
    "df = pd.merge(counts_df, df_opportunities[['Hugo_Symbol','synonymous_opportunity', 'nonsynonymous_opportunity']], how='inner', left_index=True, right_on='Hugo_Symbol')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ece3a410",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adjust nonsynonymous opportunities to include indels\n",
    "df['Indels'] = df['Frame_Shift_Del'] + df['Frame_Shift_Ins'] + df['In_Frame_Del'] + df['In_Frame_Ins']\n",
    "df['NS_SNV'] = df['Missense_Mutation'] + df['Nonsense_Mutation'] + df['Nonstop_Mutation']\n",
    "df['nonsynonymous_opportunity'] = df['nonsynonymous_opportunity'] * (1 + ((df['Indels'] + .5)/(df['NS_SNV'] + .5)).mean())\n",
    "\n",
    "df = df.set_index('Hugo_Symbol')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "1e92e252",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>synonymous</th>\n",
       "      <th>Frame_Shift_Del</th>\n",
       "      <th>Frame_Shift_Ins</th>\n",
       "      <th>In_Frame_Del</th>\n",
       "      <th>In_Frame_Ins</th>\n",
       "      <th>Missense_Mutation</th>\n",
       "      <th>Nonsense_Mutation</th>\n",
       "      <th>Nonstop_Mutation</th>\n",
       "      <th>Translation_Start_Site</th>\n",
       "      <th>non-synonymous</th>\n",
       "      <th>CDS_length</th>\n",
       "      <th>synonymous_opportunity</th>\n",
       "      <th>nonsynonymous_opportunity</th>\n",
       "      <th>Indels</th>\n",
       "      <th>NS_SNV</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1242</td>\n",
       "      <td>5003.480225</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3082</td>\n",
       "      <td>12399.823469</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3006</td>\n",
       "      <td>12273.307072</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADAC</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>805</td>\n",
       "      <td>3400.128186</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADACL4</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>855</td>\n",
       "      <td>3426.891270</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             synonymous  Frame_Shift_Del  Frame_Shift_Ins  In_Frame_Del  \\\n",
       "Hugo_Symbol                                                               \n",
       "A1CF                  1                0                0             0   \n",
       "A2M                   1                0                0             0   \n",
       "A2ML1                 4                0                0             0   \n",
       "AADAC                 0                0                0             0   \n",
       "AADACL4               1                0                0             0   \n",
       "\n",
       "             In_Frame_Ins  Missense_Mutation  Nonsense_Mutation  \\\n",
       "Hugo_Symbol                                                       \n",
       "A1CF                    0                  2                  0   \n",
       "A2M                     0                  4                  2   \n",
       "A2ML1                   0                  6                  1   \n",
       "AADAC                   0                  3                  1   \n",
       "AADACL4                 0                  2                  0   \n",
       "\n",
       "             Nonstop_Mutation  Translation_Start_Site  non-synonymous  \\\n",
       "Hugo_Symbol                                                             \n",
       "A1CF                        0                       0               2   \n",
       "A2M                         0                       0               6   \n",
       "A2ML1                       0                       0               7   \n",
       "AADAC                       0                       0               4   \n",
       "AADACL4                     0                       0               2   \n",
       "\n",
       "             CDS_length  synonymous_opportunity  nonsynonymous_opportunity  \\\n",
       "Hugo_Symbol                                                                  \n",
       "A1CF                NaN                    1242                5003.480225   \n",
       "A2M                 NaN                    3082               12399.823469   \n",
       "A2ML1               NaN                    3006               12273.307072   \n",
       "AADAC               NaN                     805                3400.128186   \n",
       "AADACL4             NaN                     855                3426.891270   \n",
       "\n",
       "             Indels  NS_SNV  \n",
       "Hugo_Symbol                  \n",
       "A1CF              0       2  \n",
       "A2M               0       6  \n",
       "A2ML1             0       7  \n",
       "AADAC             0       4  \n",
       "AADACL4           0       2  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "142f0790",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Mutation count table: SBS_count[ref][alt]\n",
    "SBS_count = defaultdict(lambda: defaultdict(int))\n",
    "\n",
    "def normalize_sbs(ref, alt):\n",
    "    # Normalize purine ref bases to pyrimidine (G→C, A→T; and likewise alt)\n",
    "    complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    if ref in ['G', 'A']:\n",
    "        ref = complement[ref]\n",
    "        alt = complement[alt]\n",
    "    return ref, alt\n",
    "\n",
    "for hgvsc in df_mut['HGVSc']:\n",
    "    m = re.search(r'c\\.\\d+([ACGT])>([ACGT])', str(hgvsc))\n",
    "    if not m:\n",
    "        continue\n",
    "    ref, alt = m.group(1), m.group(2)\n",
    "    norm_ref, norm_alt = normalize_sbs(ref, alt)\n",
    "    SBS_count[norm_ref][norm_alt] += 1\n",
    "\n",
    "records = []\n",
    "for ref in ['C', 'T']:\n",
    "    for alt in ['A', 'G', 'T'] if ref == 'C' else ['A', 'C', 'G']:\n",
    "        count = SBS_count[ref][alt]\n",
    "        records.append({'ref': ref, 'alt': alt, 'mutation': f\"{ref}>{alt}\", 'count': count})\n",
    "\n",
    "sbs_df = pd.DataFrame(records)\n",
    "\n",
    "# Count total number of 'C' and 'T' bases across all CDSs\n",
    "base_counts = defaultdict(int)\n",
    "\n",
    "for seq in df_cds['CDS sequence']:\n",
    "    seq = str(seq).upper()\n",
    "    base_counts['C'] += seq.count('C')\n",
    "    base_counts['T'] += seq.count('T')\n",
    "\n",
    "# Multiply by 2 * number of individuals (diploid genomes)\n",
    "# change num individuals to number of unique patient ids\n",
    "num_individuals = df_mut['patient_id'].nunique()\n",
    "multiplier = 2 * num_individuals\n",
    "\n",
    "# Assign opportunities for each of the 6 SBS types\n",
    "opportunities = {\n",
    "    'C>A': base_counts['C'] * multiplier,\n",
    "    'C>G': base_counts['C'] * multiplier,\n",
    "    'C>T': base_counts['C'] * multiplier,\n",
    "    'T>A': base_counts['T'] * multiplier,\n",
    "    'T>C': base_counts['T'] * multiplier,\n",
    "    'T>G': base_counts['T'] * multiplier\n",
    "}\n",
    "\n",
    "sbs_df['opportunities'] = sbs_df['mutation'].map(opportunities)\n",
    "sbs_df['rate'] = sbs_df['count'] / sbs_df['opportunities']\n",
    "\n",
    "# DNA codon table\n",
    "codon_table = {\n",
    "    'TTT': 'F', 'TTC': 'F', 'TTA': 'L', 'TTG': 'L',\n",
    "    'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S',\n",
    "    'TAT': 'Y', 'TAC': 'Y', 'TAA': '*', 'TAG': '*',\n",
    "    'TGT': 'C', 'TGC': 'C', 'TGA': '*', 'TGG': 'W',\n",
    "\n",
    "    'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L',\n",
    "    'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P',\n",
    "    'CAT': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q',\n",
    "    'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R',\n",
    "\n",
    "    'ATT': 'I', 'ATC': 'I', 'ATA': 'I', 'ATG': 'M',\n",
    "    'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T',\n",
    "    'AAT': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K',\n",
    "    'AGT': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R',\n",
    "\n",
    "    'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V',\n",
    "    'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A',\n",
    "    'GAT': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E',\n",
    "    'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'\n",
    "}\n",
    "\n",
    "mutation_rate_map = sbs_df.set_index('mutation')['rate'].to_dict()\n",
    "\n",
    "codon_opportunity = {}\n",
    "\n",
    "for codon in codon_table:\n",
    "    syn = 0.0\n",
    "    nonsyn = 0.0\n",
    "    ref_aa = codon_table[codon]\n",
    "\n",
    "    for pos in range(3):\n",
    "        ref_base = codon[pos]\n",
    "        for alt_base in \"ACGT\":\n",
    "            if alt_base == ref_base:\n",
    "                continue\n",
    "            mut_codon = codon[:pos] + alt_base + codon[pos+1:]\n",
    "            if mut_codon not in codon_table:\n",
    "                continue\n",
    "            alt_aa = codon_table[mut_codon]\n",
    "\n",
    "            norm_ref, norm_alt = normalize_sbs(ref_base, alt_base)\n",
    "            key = f\"{norm_ref}>{norm_alt}\"\n",
    "            rate = mutation_rate_map.get(key, 0.0)\n",
    "\n",
    "            if ref_aa == alt_aa:\n",
    "                syn += rate\n",
    "            else:\n",
    "                nonsyn += rate\n",
    "\n",
    "    codon_opportunity[codon] = (syn, nonsyn)\n",
    "\n",
    "results = []\n",
    "\n",
    "for idx, row in df_cds.iterrows():\n",
    "    gene = row['Hugo_Symbol']\n",
    "    seq = row['CDS sequence'].upper()\n",
    "\n",
    "    if gene not in df.index:\n",
    "        continue  # skip genes without observed mutations\n",
    "\n",
    "    O_NS = df.loc[gene, 'non-synonymous']\n",
    "    O_S = df.loc[gene, 'synonymous']\n",
    "\n",
    "    E_S, E_NS = 0.0, 0.0\n",
    "    for i in range(0, len(seq) - 2, 3):\n",
    "        codon = seq[i:i+3]\n",
    "        if codon not in codon_opportunity:\n",
    "            continue\n",
    "        syn, nonsyn = codon_opportunity[codon]\n",
    "        E_S += syn\n",
    "        E_NS += nonsyn\n",
    "\n",
    "    E_S = E_S * 2 * num_individuals\n",
    "    E_NS = E_NS * 2 * num_individuals\n",
    "\n",
    "    λ = O_S / E_S if E_S > 0 else np.nan\n",
    "    expected_NS_scaled = λ * E_NS if λ is not None else None\n",
    "    dnds = O_NS / expected_NS_scaled if expected_NS_scaled else np.nan\n",
    "\n",
    "    results.append({\n",
    "        'Hugo_Symbol': gene,\n",
    "        'synonymous': O_S,\n",
    "        'observed_nonsynonymous': O_NS,\n",
    "        'synonymous_opportunity': E_S,\n",
    "        'nonsynonymous_opportunity': E_NS,\n",
    "        'λ': λ,\n",
    "        'dN/dS': dnds\n",
    "    })\n",
    "\n",
    "df_dnds = pd.DataFrame(results)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "c87017e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>observed_nonsynonymous</th>\n",
       "      <th>synonymous_opportunity</th>\n",
       "      <th>nonsynonymous_opportunity</th>\n",
       "      <th>λ</th>\n",
       "      <th>dN/dS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5368</th>\n",
       "      <td>TP53</td>\n",
       "      <td>2</td>\n",
       "      <td>246</td>\n",
       "      <td>1.239781</td>\n",
       "      <td>3.097287</td>\n",
       "      <td>1.613188</td>\n",
       "      <td>49.234399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5064</th>\n",
       "      <td>GATA3</td>\n",
       "      <td>2</td>\n",
       "      <td>75</td>\n",
       "      <td>1.755706</td>\n",
       "      <td>3.530890</td>\n",
       "      <td>1.139143</td>\n",
       "      <td>18.646561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1736</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>6</td>\n",
       "      <td>291</td>\n",
       "      <td>2.157271</td>\n",
       "      <td>7.328571</td>\n",
       "      <td>2.781292</td>\n",
       "      <td>14.276673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4884</th>\n",
       "      <td>FOXA1</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>1.887725</td>\n",
       "      <td>3.852090</td>\n",
       "      <td>0.529738</td>\n",
       "      <td>12.251301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2443</th>\n",
       "      <td>CDH1</td>\n",
       "      <td>3</td>\n",
       "      <td>88</td>\n",
       "      <td>2.675803</td>\n",
       "      <td>6.475851</td>\n",
       "      <td>1.121159</td>\n",
       "      <td>12.120449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>831</th>\n",
       "      <td>MAP3K1</td>\n",
       "      <td>3</td>\n",
       "      <td>89</td>\n",
       "      <td>3.749497</td>\n",
       "      <td>11.402825</td>\n",
       "      <td>0.800107</td>\n",
       "      <td>9.755046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5337</th>\n",
       "      <td>MAP2K4</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>0.953652</td>\n",
       "      <td>2.888429</td>\n",
       "      <td>1.048601</td>\n",
       "      <td>8.254070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>ARID1A</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>7.259698</td>\n",
       "      <td>18.720393</td>\n",
       "      <td>0.137747</td>\n",
       "      <td>8.143721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>LRP2</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>11.323221</td>\n",
       "      <td>34.319247</td>\n",
       "      <td>0.088314</td>\n",
       "      <td>7.918510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4802</th>\n",
       "      <td>RUNX1</td>\n",
       "      <td>2</td>\n",
       "      <td>31</td>\n",
       "      <td>1.801039</td>\n",
       "      <td>3.840226</td>\n",
       "      <td>1.110470</td>\n",
       "      <td>7.269389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Hugo_Symbol  synonymous  observed_nonsynonymous  synonymous_opportunity  \\\n",
       "5368        TP53           2                     246                1.239781   \n",
       "5064       GATA3           2                      75                1.755706   \n",
       "1736      PIK3CA           6                     291                2.157271   \n",
       "4884       FOXA1           1                      25                1.887725   \n",
       "2443        CDH1           3                      88                2.675803   \n",
       "831       MAP3K1           3                      89                3.749497   \n",
       "5337      MAP2K4           1                      25                0.953652   \n",
       "304       ARID1A           1                      21                7.259698   \n",
       "36          LRP2           1                      24               11.323221   \n",
       "4802       RUNX1           2                      31                1.801039   \n",
       "\n",
       "      nonsynonymous_opportunity         λ      dN/dS  \n",
       "5368                   3.097287  1.613188  49.234399  \n",
       "5064                   3.530890  1.139143  18.646561  \n",
       "1736                   7.328571  2.781292  14.276673  \n",
       "4884                   3.852090  0.529738  12.251301  \n",
       "2443                   6.475851  1.121159  12.120449  \n",
       "831                   11.402825  0.800107   9.755046  \n",
       "5337                   2.888429  1.048601   8.254070  \n",
       "304                   18.720393  0.137747   8.143721  \n",
       "36                    34.319247  0.088314   7.918510  \n",
       "4802                   3.840226  1.110470   7.269389  "
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dnds.sort_values(by='dN/dS', ascending=False, inplace=True)\n",
    "df_dnds.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "526c69f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def poisson_pval(x):\n",
    "    \"\"\"P-value for observed nonsynonymous mutation count of the gene. Under null, expected should be equal to the observed synonymous mutations times the ratio of nonsynonymous to synonymous opportunities.\"\"\"\n",
    "    return 1 - poisson.cdf(x['observed_nonsynonymous']-1, x['synonymous']*x['nonsynonymous_opportunity']/x['synonymous_opportunity'])\n",
    "\n",
    "df_dnds['poisson_pval'] = df_dnds.apply(lambda x: poisson_pval(x), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "5438001e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using poisson_pval, DCG: 0.5202934952448164, Bpref: 0.17296786389413987, Accuracy: 0.17391304347826086\n"
     ]
    }
   ],
   "source": [
    "from src.utils.eval  import evalAccuracy, compareRankings\n",
    "\n",
    "# load the IntOGen ranking TSV\n",
    "path_intogen = \"data/raw/IntOGen-DriverGenes_TCGA_WXS_BRCA.tsv\"\n",
    "df_intogen = pd.read_csv(path_intogen, sep=\"\\t\")\n",
    "\n",
    "# build a dict mapping gene names to their IntOGen relevance.\n",
    "#RELEVANCE IS SAMPLES%\n",
    "baseline_ranks = dict(zip(df_intogen[\"Symbol\"], (df_intogen[\"Samples (%)\"] * 0.01)))\n",
    "\n",
    "df_dn_ds = df_dnds\n",
    "#df_dn_df = df_dn_ds['Hugo_Symbol'].dropna()\n",
    "#dn_ds_ranks = dict(zip(df_dn_ds[\"Hugo_Symbol\"], df_dn_ds[\"dN/dS\"]))\n",
    "\n",
    "dn_ds_ranks = dict(zip(df_dn_ds[\"Hugo_Symbol\"], 1- df_dn_ds[\"poisson_pval\"]))\n",
    "#dn_ds_ranks = dict(zip(df_dn_ds[\"Hugo_Symbol\"], df_dn_ds[\"Hugo_Symbol\"]))\n",
    "\n",
    "# calculate accuracy metrics\n",
    "dcg, bpref, accuracy = evalAccuracy(dn_ds_ranks, baseline_ranks)\n",
    "\n",
    "df_gene_ranks = compareRankings(dn_ds_ranks, baseline_ranks)\n",
    "\n",
    "print(f\"Using poisson_pval, DCG: {dcg}, Bpref: {bpref}, Accuracy: {accuracy}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c841ff83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using dN/dS, DCG: 0.5335676792849361, Bpref: 0.24196597353497168, Accuracy: 0.26086956521739135\n"
     ]
    }
   ],
   "source": [
    "# load the IntOGen ranking TSV\n",
    "path_intogen = \"data/raw/IntOGen-DriverGenes_TCGA_WXS_BRCA.tsv\"\n",
    "df_intogen = pd.read_csv(path_intogen, sep=\"\\t\")\n",
    "\n",
    "# build a dict mapping gene names to their IntOGen relevance.\n",
    "#RELEVANCE IS SAMPLES%\n",
    "baseline_ranks = dict(zip(df_intogen[\"Symbol\"], (df_intogen[\"Samples (%)\"] * 0.01)))\n",
    "\n",
    "df_dn_ds = df_dnds\n",
    "df_dn_df = df_dn_ds[df_dn_ds['Hugo_Symbol'].notna()]\n",
    "\n",
    "dn_ds_ranks = dict(zip(df_dn_ds[\"Hugo_Symbol\"], df_dn_ds[\"dN/dS\"]))\n",
    "\n",
    "# calculate accuracy metrics\n",
    "dcg, bpref, accuracy = evalAccuracy(dn_ds_ranks, baseline_ranks)\n",
    "\n",
    "df_gene_ranks = compareRankings(dn_ds_ranks, baseline_ranks)\n",
    "\n",
    "print(f\"Using dN/dS, DCG: {dcg}, Bpref: {bpref}, Accuracy: {accuracy}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da00464c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "project2_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
